docs/rfc/0698-cli-infrastructure.adoc

```diff
@@ -0,0 +1,299 @@
+maps, where the map key is the zero-based numeric index of the field in
+declaration order.
+
+== One `rad` to Rule Them All
+
+[quote]
+-------
+One ring to rule them all, one ring to find them,
+One ring to bring them all, and in the darkness bind them;
+-------
+-- J.R.R. Tolkein, Fellowship of the Ring
+
+The integral part of this proposal is that there will be one binary,
+`rad`, that will be the foundation for all other subcommands. It will
+define the global parameters and how to extend it.
+
+We will discuss these in the following two sections.
+
+== The Fellowship of the `rad`
+
+[quoute]
+--------
+“I can’t carry it for you, but I can carry you.” – Sam Gamgee
+--------
+-- J.R.R. Tolkien, The Return of the King
+
+To allow the ecosystem of commands to be extended by ourselves and
+others in the community it will be necessary to extend the `rad`
+command with further subcommands. The details of this are outlined in
+<<arch-cli, RFC-682#CLI>>, however, we will give a quick overview here for the sake
+of the reader.
+
+Taking inspiration from tools such as `git` and `cargo`, we propose
+that the `rad` binary can be extended by placing another
+binary on the users `$PATH` (or `%PATH%` on Windows). This
+binary MUST follow the naming convention of `rad-*`. To reiterate
+here, these new commands MUST NOT override the core set outlined here
+(nor in further RFCs).
+
+== Global Parameters
+
+Many subcommands will likely want to share a set of options. These
+options, dubbed as "global parameters" here, will be a defined set --
+exported from `radicle-link` for re-use -- that the base `rad` command
+will always receive and pass onto the subcommand being called.
+
+Other subcommands that are written in the target language (currently
+Rust) will be able to import the `struct` for reuse. If the subcommand
+is in another target language it SHALL add these parameters to its
+parameter list prefixed with `global-`. The `rad` command will parse
+these parameters, setting defaults if necessary, and SHALL pass them
+varbatim to the subcommand.
+
+Thus, all subcommands SHALL receive these global parameters as part of
+their interface. We list these parameters here, but it is NOT an
+exhaustive list:
+
+* `global-verbose`
+* `global-quiet`
+* `global-profile := active (default) | <profile-id>`
+
+== Online & Offline
+
+In <<arch-orch, RFC-682#Process-Orchestration>>, it was decided that the peer-to-peer
+damon would not be a long-running process. The terminology we will use
+here is that the daemon is *online* if it is running and connected to
+the network. Otherwise, if it is not running then it is *offline*.
+
+This disctinction becomes important when thinking about the
+functionality of `rad` subcommands. The majority of subcommands that
+interact with the storage can be done while the daemon is offline. The
+only action the author can currently think of — that is online — is
+replication of a URN from the network.
```

Comment:
I think this was highlighted in another thread - which seems to be lost now in GHs diff view :shrug: - #114 would allow for all operations being "offline-first" and could be alleviate as a blocking task to realise this grand CLI/daemon vision.

Reply:
Nonetheless the replication _itself_ requires you need to be connected to the network. The `ask` functionality outlined below will need #141. Perhaps this section can be revised. I first wrote it when I thought the PubSub was tied to the daemon being online, which turned out to be a misunderstanding on my part :)

Reply:
A revision of the section would be helpful, as it seems we are converging on a model - given the prerequisites are in place - where there is no interactive protocol in place to achieve tasks like "asking"/tracking.